app/models/service/resource_linking.rb

```diff
@@ -0,0 +1,37 @@
+module Service::ResourceLinking
+  extend ActiveSupport::Concern
+
+  def add_provider_vms_queue(provider, array_of_uid_ems)
+    provider_id = provider.kind_of?(ExtManagementSystem) ? provider.id : provider
+    task_opts = {
+      :action => "linking VMs to service for user #{evm_owner.userid}",
+      :userid => evm_owner.userid
+    }
+    queue_opts = {
+      :class_name  => "Service",
+      :method_name => "add_provider_vms",
+      :instance_id => id,
+      :zone        => my_zone,
+      :args        => [provider_id, array_of_uid_ems]
+    }
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def add_provider_vms(provider, array_of_uid_ems)
+    provider = ExtManagementSystem.find(provider) if provider.kind_of?(Integer)
+    array_of_uid_ems = array_of_uid_ems.kind_of?(Array) ? array_of_uid_ems.compact.uniq : [array_of_uid_ems]
+
+    if Vm.where(:uid_ems => array_of_uid_ems).length != array_of_uid_ems.length
+      task_ids = EmsRefresh.queue_refresh(provider, nil, :create_task => true)
+
+      # Wait for the tasks to finish
+      task_ids.each { |tid| MiqTask.wait_for_taskid(tid) }
```

Comment:
We should not be waiting for the provider refresh synchronously, instead it should be re-queueing the work and checking.  Some providers could take more then 10 mins to complete a refresh.  This also means we cannot use `MiqTask.generic_action_with_callback`.

cc @mkanoor